File: backend/app/routers/presentations.py
from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, BackgroundTasks, Body, Depends, File, Form, HTTPException, UploadFile, status
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.core.content_disposition import attachment_content_disposition
from app.core.deps import get_current_user
from app.database import get_db
from app.models import User
from app.schemas.presentations import (
    PresentationBuildRequest,
    PresentationBuildResponse,
    PresentationCreateResponse,
    PresentationOutlineUpdateRequest,
    PresentationOutlineUpdateResponse,
    PresentationStatusResponse,
)
from app.services import presentation_build_service as build_service
from app.services import presentation_service as presentations

router = APIRouter(prefix="/api/presentations", tags=["presentations"])


def _slide_count(presentation) -> Optional[int]:
    if not presentation.slides_json:
        return None
    slides = presentation.slides_json.get("slides")
    if isinstance(slides, list):
        return len(slides)
    return None


@router.post("/create", response_model=PresentationCreateResponse)
async def create_presentation(
    prompt: str = Form(default=""),
    agent_id: Optional[str] = Form(default=None),
    template_id: Optional[int] = Form(default=None),
    source_files: List[UploadFile] = File(default=[]),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    uploads = [item for item in source_files if item.filename]
    result = await presentations.create_presentation_outline(
        db,
        user=current_user,
        prompt=prompt,
        agent_id=agent_id,
        template_id=template_id,
        source_files=uploads,
    )
    return PresentationCreateResponse(**result)


@router.patch("/{presentation_id}/outline", response_model=PresentationOutlineUpdateResponse)
async def update_presentation_outline(
    presentation_id: int,
    payload: PresentationOutlineUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    presentation = presentations.update_presentation_outline(
        db,
        user=current_user,
        presentation_id=presentation_id,
        outline=payload.outline,
    )
    return PresentationOutlineUpdateResponse(
        id=presentation.id,
        outline=presentation.outline,
        title=presentation.title,
        status=presentation.status,
    )


@router.post("/{presentation_id}/build", response_model=PresentationBuildResponse)
async def build_presentation(
    presentation_id: int,
    background_tasks: BackgroundTasks,
    payload: PresentationBuildRequest = Body(default_factory=PresentationBuildRequest),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    presentation = build_service.get_user_presentation(
        db,
        user_id=current_user.id,
        presentation_id=presentation_id,
    )
    if presentation.template_id is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Выберите шаблон перед сборкой презентации",
        )
    if presentation.status == "building":
        return PresentationBuildResponse(
            id=presentation.id,
            status=presentation.status,
            error_message=None,
            slide_count=_slide_count(presentation),
        )

    presentation.status = "building"
    presentation.error_message = None
    presentation.build_stage = "queued"
    db.commit()

    background_tasks.add_task(
        build_service.run_build_in_background,
        presentation_id=presentation_id,
        user_id=current_user.id,
        generate_images=payload.generate_images,
    )

    return PresentationBuildResponse(
        id=presentation.id,
        status="building",
        error_message=None,
        slide_count=_slide_count(presentation),
    )


@router.get("/{presentation_id}", response_model=PresentationStatusResponse)
async def get_presentation_status(
    presentation_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    presentation = build_service.get_user_presentation(
        db,
        user_id=current_user.id,
        presentation_id=presentation_id,
    )
    return PresentationStatusResponse(
        id=presentation.id,
        status=presentation.status,
        title=presentation.title,
        outline=presentation.outline,
        template_id=presentation.template_id,
        build_stage=presentation.build_stage,
        error_message=presentation.error_message,
        slide_count=_slide_count(presentation),
        has_download=bool(presentation.pptx_data),
        created_at=presentation.created_at,
        updated_at=presentation.updated_at,
    )


@router.get("/{presentation_id}/download")
async def download_presentation(
    presentation_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    presentation = build_service.get_user_presentation(
        db,
        user_id=current_user.id,
        presentation_id=presentation_id,
    )
    if presentation.status != "ready" or not presentation.pptx_data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Файл презентации ещё не готов",
        )

    filename = f"{presentation.title.strip() or 'presentation'}.pptx"

    return Response(
        content=presentation.pptx_data,
        media_type="application/vnd.openxmlformats-officedocument.presentationml.presentation",
        headers={"Content-Disposition": attachment_content_disposition(filename)},
    )

File: frontend/lib/api/presentations.ts
import { ApiError } from "@/lib/api/auth";
import { getAccessToken } from "@/lib/auth/token";
import type { AgentId } from "@/lib/api/agents";

const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

function parseApiError(data: unknown, fallback: string): string {
  if (!data || typeof data !== "object") {
    return fallback;
  }
  const detail = (data as { detail?: unknown }).detail;
  if (typeof detail === "string") {
    return detail;
  }
  return fallback;
}

function authHeaders(): HeadersInit {
  const token = getAccessToken();
  if (!token) {
    throw new ApiError("Требуется авторизация", 401);
  }
  return { Authorization: `Bearer ${token}` };
}

export type PresentationStatus = "draft" | "building" | "ready" | "failed";

export type PresentationCreateResponse = {
  id: number;
  outline: string;
  agent_id: AgentId;
  template_id: number | null;
  template_name: string | null;
  template_file_type: "pptx" | "pdf" | null;
  source_files: string[];
  status: PresentationStatus;
};

export type PresentationBuildResponse = {
  id: number;
  status: PresentationStatus;
  error_message: string | null;
  slide_count: number | null;
};

export type PresentationStatusResponse = {
  id: number;
  status: PresentationStatus;
  title: string;
  outline: string | null;
  template_id: number | null;
  template_name: string | null;
  template_file_type: "pptx" | "pdf" | null;
  build_stage: string | null;
  error_message: string | null;
  slide_count: number | null;
  has_download: boolean;
  created_at: string;
  updated_at: string;
};

export type PresentationListItem = {
  id: number;
  title: string;
  status: PresentationStatus;
  template_id: number | null;
  template_name: string | null;
  template_file_type: "pptx" | "pdf" | null;
  slide_count: number | null;
  has_download: boolean;
  created_at: string;
  updated_at: string;
};

export type PresentationOutlineUpdateResponse = {
  id: number;
  outline: string;
  title: string;
  status: PresentationStatus;
};

async function apiAuthFetch<T>(path: string, init: RequestInit = {}): Promise<T> {
  const headers = new Headers(init.headers);
  const auth = authHeaders();
  Object.entries(auth).forEach(([key, value]) => headers.set(key, value));

  const response = await fetch(`${API_URL}${path}`, { ...init, headers });

  if (!response.ok) {
    const errorData = await response.json().catch(() => null);
    throw new ApiError(parseApiError(errorData, "Ошибка запроса"), response.status);
  }

  return (await response.json()) as T;
}

export async function fetchPresentations(): Promise<PresentationListItem[]> {
  return apiAuthFetch<PresentationListItem[]>("/api/presentations");
}

export async function createPresentation(payload: {
  prompt: string;
  agentId: AgentId;
  templateId: number | null;
  sourceFiles: File[];
}): Promise<PresentationCreateResponse> {
  const formData = new FormData();
  formData.append("prompt", payload.prompt);
  formData.append("agent_id", payload.agentId);
  if (payload.templateId !== null) {
    formData.append("template_id", String(payload.templateId));
  }
  for (const file of payload.sourceFiles) {
    formData.append("source_files", file);
  }

  const response = await fetch(`${API_URL}/api/presentations/create`, {
    method: "POST",
    headers: authHeaders(),
    body: formData,
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => null);
    throw new ApiError(parseApiError(errorData, "Не удалось создать презентацию"), response.status);
  }

  return (await response.json()) as PresentationCreateResponse;
}

export async function updatePresentationOutline(
  presentationId: number,
  outline: string,
): Promise<PresentationOutlineUpdateResponse> {
  const response = await fetch(`${API_URL}/api/presentations/${presentationId}/outline`, {
    method: "PATCH",
    headers: {
      ...authHeaders(),
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ outline }),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => null);
    throw new ApiError(parseApiError(errorData, "Не удалось сохранить план"), response.status);
  }

  return (await response.json()) as PresentationOutlineUpdateResponse;
}

export async function renamePresentation(
  presentationId: number,
  title: string,
): Promise<PresentationListItem> {
  const data = await apiAuthFetch<{ presentation: PresentationListItem }>(
    `/api/presentations/${presentationId}`,
    {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ title: title.trim() }),
    },
  );
  return data.presentation;
}

export async function deletePresentation(presentationId: number): Promise<void> {
  await apiAuthFetch<{ message: string }>(`/api/presentations/${presentationId}`, {
    method: "DELETE",
  });
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Запускает сборку (сразу возвращает building) и ждёт ready/failed через polling. */
export async function buildPresentation(
  presentationId: number,
  options?: {
    generateImages?: boolean;
    onPoll?: (status: PresentationStatusResponse) => void;
  },
): Promise<PresentationBuildResponse> {
  const generateImages = options?.generateImages ?? true;
  const onPoll = options?.onPoll;
  const response = await fetch(`${API_URL}/api/presentations/${presentationId}/build`, {
    method: "POST",
    headers: {
      ...authHeaders(),
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ generate_images: generateImages }),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => null);
    throw new ApiError(parseApiError(errorData, "Не удалось собрать презентацию"), response.status);
  }

  const started = (await response.json()) as PresentationBuildResponse;
  if (started.status !== "building") {
    return started;
  }

  const deadline = Date.now() + 20 * 60 * 1000;
  while (Date.now() < deadline) {
    await sleep(3000);
    const status = await getPresentationStatus(presentationId);
    onPoll?.(status);
    if (status.status === "ready") {
      return {
        id: status.id,
        status: "ready",
        error_message: null,
        slide_count: status.slide_count,
      };
    }
    if (status.status === "failed") {
      return {
        id: status.id,
        status: "failed",
        error_message: status.error_message,
        slide_count: status.slide_count,
      };
    }
  }

  throw new ApiError(
    "Сборка занимает слишком много времени. Проверьте логи backend или повторите позже.",
    504,
  );
}

export async function getPresentationStatus(
  presentationId: number,
): Promise<PresentationStatusResponse> {
  const response = await fetch(`${API_URL}/api/presentations/${presentationId}`, {
    headers: authHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => null);
    throw new ApiError(parseApiError(errorData, "Не удалось получить статус"), response.status);
  }

  return (await response.json()) as PresentationStatusResponse;
}

export async function downloadPresentation(presentationId: number, title: string): Promise<void> {
  const response = await fetch(`${API_URL}/api/presentations/${presentationId}/download`, {
    headers: authHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => null);
    throw new ApiError(parseApiError(errorData, "Не удалось скачать файл"), response.status);
  }

  const blob = await response.blob();
  const safeTitle = title.trim() || "presentation";
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = `${safeTitle}.pptx`;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}
